src/webapp/cilogon_ldap.py

```diff
@@ -0,0 +1,44 @@
+import ldap3
+
+
+def get_contact_cilogon_id_map(global_data):
+    """ return contacts dict, limited to users with a CILogonID """
+    contacts = global_data.get_contacts_data().users_by_id;
+    return { k: v for k, v in contacts.items() if v.cilogon_id is not None }
+
+
+# cilogon ldap query constants
+_ldap_url = "ldaps://ldap.cilogon.org"
+_username = "uid=readonly_user,ou=system,o=OSG,o=CO,dc=cilogon,dc=org"
+_basedn   = "o=OSG,o=CO,dc=cilogon,dc=org"
+
+
+def get_cilogon_ldap_id_map(ldappass):
+    """ return dict of cilogon ldap data for each CILogonID, with the
+        structure: {CILogonID: { "dn": dn, "data": data }, ...} """
+    server = ldap3.Server(_ldap_url)
+    conn = ldap3.Connection(server, _username, ldappass)
+    if not conn.bind():
+        return None  # connection failure
+    conn.search(_basedn, '(voPersonID=*)', attributes=['*'])
+    result_data = [ (e.entry_dn, e.entry_attributes_as_dict)
+                    for e in conn.entries ]
```

Comment:
Is the `entry_dn` really a separate attr from the rest of `entry_attributes_as_dict`? What sort of info are we grabbing from `entry_attributes_as_dict`?

Reply:
Yeah so, `conn.entries` here is a list of `ldap3.abstract.entry.Entry` objects returned by the `conn.search(...)` call.

Each of these entries has a dn and an "attributes", which can be accessed as a dict.  For instance calling `e.entry_to_json()` on a single entry results in

```
{
    "attributes": {
        "cn": [
            "Brian Bockelman"
        ],
        "eduPersonOrcid": [
            "0000-0003-2981-3809"
        ],
        "eduPersonPrincipalName": [
            "bockelman@wisc.edu"
        ],
        "givenName": [
            "Brian"
        ],
        "isMemberOf": [
            "CO:members:all",
            "CO:members:active",
            "CO:admins",
            "hub-admins",
            "CO:COU:SOTERIA:members:active",
            "CO:COU:SOTERIA:members:all",
            "soteria-brian-bockelman"
        ],
        "mail": [
            "BBOCKELMAN@MORGRIDGE.ORG"
        ],
        "objectClass": [
            "person",
            "organizationalPerson",
            "inetOrgPerson",
            "eduPerson",
            "eduMember",
            "voPerson"
        ],
        "sn": [
            "Bockelman"
        ],
        "uid": [
            "http://cilogon.org/serverA/users/9265706"
        ],
        "voPersonID": [
            "OSG1000001"
        ]
    },
    "dn": "voPersonID=OSG1000001,ou=people,o=OSG,o=CO,dc=cilogon,dc=org"
}
```

...

> What sort of info are we grabbing from `entry_attributes_as_dict`?

This function returns a map of `voPersonID: { "dn": dn, "data": data }` for each entry, where `data` is the attributes dict.  (The voPersonID is the CILogonID.)

I suspect i included the "dn" from the entry here because it was not otherwise present in the "attributes" data.  Though as far as i can tell we don't actually use the returned dn here.

Separately (and maybe the only part that's currently used), `cilogon_id_map_to_ssh_keys` takes that returned map and grabs the `sshPublicKey` attr, transforming it into a map of `voPersonID: [sshPublicKey, ...]`.  This is used in oasis_managers to get the ssh public keys for vo managers.

Reply:
Ah ok, that makes sense. I was curious about why the DN was separate but this seems like a fine way to organize things